pkg/http/sender.go

```diff
@@ -0,0 +1,660 @@
+		case "min_throughput", "init_buf_size":
+			parsedVal, err := strconv.Atoi(v)
+			if err != nil {
+				return nil, conf.NewConfigStrParseError("invalid %s value, %q is not a valid int", k, v)
+
+			}
+			switch k {
+			case "min_throughput":
+				opts = append(opts, WithMinThroughput(parsedVal))
+			case "init_buf_size":
+				opts = append(opts, WithInitBufferSize(parsedVal))
+			default:
+				panic("add a case for " + k)
+			}
+
+		case "request_timeout", "retry_timeout":
+			timeout, err := strconv.Atoi(v)
+			if err != nil {
+				return nil, conf.NewConfigStrParseError("invalid %s value, %q is not a valid int", k, v)
+			}
+
+			timeoutDur := time.Duration(timeout * int(time.Millisecond))
+
+			switch k {
+			case "request_timeout":
+				opts = append(opts, WithRequestTimeout(timeoutDur))
+			case "retry_timeout":
+				opts = append(opts, WithRetryTimeout(timeoutDur))
+			default:
+				panic("add a case for " + k)
+			}
+		case "tls_verify":
+			switch v {
+			case "on":
+				opts = append(opts, WithTls())
+			case "unsafe_off":
+				opts = append(opts, WithTlsInsecureSkipVerify())
+			default:
+				return nil, conf.NewConfigStrParseError("invalid tls_verify value, %q is not 'on' or 'unsafe_off", v)
+			}
+		case "tls_roots":
+			return nil, conf.NewConfigStrParseError("tls_roots is not available in the go client")
+		case "tls_roots_password":
+			return nil, conf.NewConfigStrParseError("tls_roots_password is not available in the go client")
+		default:
+			return nil, conf.NewConfigStrParseError("unsupported option %q", k)
+		}
+
+	}
+	return NewLineSender(opts...)
+}
+
+// Flush flushes the accumulated messages to the underlying HTTP
+// client. Should be called periodically to make sure that
+// all messages are sent to the server.
+//
+// For optimal performance, this method should not be called after
+// each ILP message. Instead, the messages should be written in
+// batches followed by a Flush call. Optimal batch size may vary
+// from 100 to 1,000 messages depending on the message size and
+// configured buffer capacity.
+func (s *LineSender) Flush(ctx context.Context) error {
+	var (
+		req           *http.Request
+		retryInterval time.Duration
+
+		maxRetryInterval = time.Second
+	)
+
+	if s.closed {
+		return errors.New("cannot flush a closed LineSender")
+	}
+
+	err := s.LastErr()
+	s.ClearLastErr()
+	if err != nil {
+		s.DiscardPendingMsg()
+		return err
+	}
+	if s.HasTable() {
+		s.DiscardPendingMsg()
+		return errors.New("pending ILP message must be finalized with At or AtNow before calling Flush")
+	}
+
+	// timeout = ( request.len() / min_throughput ) + request_timeout
+	// nb: conversion from int to time.Duration is in milliseconds
+	timeout := time.Duration(s.Len()/s.minThroughputBytesPerSecond)*time.Second + s.requestTimeout
+	reqCtx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+
+	req, err = http.NewRequestWithContext(
+		reqCtx,
+		http.MethodPost,
+		s.uri,
+		s,
```

Comment:
The buffer is set in the `Buffer.WriteTo` method that's called by the http client.  I added a test confirming this behavior

Reply:
I see. It looks like an internal standard library's fast path since they haven't documented it properly:
https://github.com/golang/go/blob/a46ecdca36b2359e7496c6d694703a9a9706d760/src/net/http/client.go#L706
https://github.com/golang/go/blob/a46ecdca36b2359e7496c6d694703a9a9706d760/src/io/io.go#L408-L412

If std lib stops checking `WriterTo` interface in future, our sender will break. Worth at least adding a clarifying comment.